consensus/src/simplex/actors/voter/actor.rs

```diff
@@ -528,560 +198,285 @@ impl<
-                );
-                resolver.fetch(vec![view], vec![view]).await;
+    /// Attempt to verify a proposed block.
+    async fn try_verify(&mut self) -> Option<(Context<D, P>, oneshot::Receiver<bool>)> {
+        // Check if we are ready to verify
+        let current_view = self.state.current_view();
+        let (context, proposal) = match self.state.try_verify(current_view) {
+            VerifyStatus::Ready(ready) => (ready.context, ready.proposal),
+            VerifyStatus::NotReady => {
                 return None;
             }
         };
+        // Unlike during proposal, we don't use a verification opportunity
+        // to backfill missing certificates (a malicious proposer could
+        // ask us to fetch junk).
 
-        // Request proposal from application
-        debug!(view = self.view, "requested proposal from automaton");
-        let context = Context {
-            round: Rnd::new(self.epoch, self.view),
-            leader: self
-                .scheme
-                .participants()
-                .key(leader)
-                .expect("leader not found")
-                .clone(),
-            parent: (parent_view, parent_payload),
-        };
-        Some((context.clone(), self.automaton.propose(context).await))
+        // Request verification
+        debug!(?proposal, "requested proposal verification");
+        Some((
+            context.clone(),
+            self.automaton.verify(context, proposal.payload).await,
+        ))
     }
 
-    fn timeout_deadline(&mut self) -> SystemTime {
-        // Return the earliest deadline
-        let view = self.views.get_mut(&self.view).unwrap();
-        if let Some(deadline) = view.leader_deadline {
-            return deadline;
-        }
-        if let Some(deadline) = view.advance_deadline {
-            return deadline;
-        }
-
-        // If no deadlines are still set (waiting for nullify),
-        // return next try for nullify.
-        if let Some(deadline) = view.nullify_retry {
-            return deadline;
+    /// Store a newly verified block.
+    async fn verified(&mut self, view: View) -> bool {
+        let round = Rnd::new(self.state.epoch(), view);
+        match self.state.verified(view) {
+            Some(Ok(proposal)) => {
+                debug!(?proposal, "successful verification");
+                true
+            }
+            Some(Err(HandleError::TimedOut)) => {
+                debug!(
+                    ?round,
+                    reason = "view timed out",
+                    "dropping verified proposal"
+                );
+                false
+            }
+            Some(Err(HandleError::NotPending)) => {
+                debug!(?round, reason = "not pending", "dropping verified proposal");
+                false
+            }
+            None => {
+                // View is no longer relevant, drop the verified proposal
+                false
+            }
         }
```

Comment:
The ergonomics could be much better here if you:
- Have `verified` return a `Result<>` rather than a `Option<Result<>>`
- Make `HandleError` a true `thiserror` type with debug

That way you can just use `?err` in the debug rather than having this have so many different cases